packages/puppeteer-core/src/api/Locator.ts

```diff
@@ -0,0 +1,374 @@
+/**
+ * Copyright 2023 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {AbortError, TimeoutError} from '../common/Errors.js';
+import {EventEmitter} from '../common/EventEmitter.js';
+import type {MouseButton} from '../common/Input.js';
+import {debugError} from '../common/util.js';
+import {isErrorLike} from '../util/ErrorLike.js';
+
+import {ElementHandle, BoundingBox} from './ElementHandle.js';
+import type {Page} from './Page.js';
+
+/**
+ * @internal
+ */
+export interface LocatorOptions {
+  /**
+   * Whether to wait for the element to be `visible` or `hidden`.
+   */
+  visibility: 'hidden' | 'visible';
+  /**
+   * Total timeout for the entire locator operation.
+   */
+  timeout: number;
+  /**
+   * Timeout for individual operations inside the locator. On errors the
+   * operation is retried as long as {@link LocatorOptions.timeout} is not
+   * exceeded. This timeout should be generally much lower as locating an
+   * element means multiple asynchronious operations.
+   */
+  operationTimeout: number;
```

Comment:
We should remove this in favor of just `timeout`. If a specific operation needs to be timed out, then there should be a `timeout` on that specific operation.

Reply:
Could elaborate why?

Reply:
My reasoning: we cannot have a single timeout setting. With a single timeout setting, we either make the overall waiting time unpredictable or we only ever run a single waitFor* without any retries because the first timeout would cancel the entire operation. I believe we need shorter timeout for the operations we retry. The constructor will be internal so it is not exposed to the users.

Reply:
Consider the following scenario (let's assume we only to waitForSelector + waitForVisible). The user loads a page and tells us to click the visible element matching the selector. So the page is loading and the first element to be rendered in the DOM is matching the selector and is not visible. If we have used only a single timeout setting, that would be the end of the locator, we would throw an error because the element we grabbed first will never become visible. With smaller timeouts for individual waitFor operations, we would time out and retry waitForSelector + waitForVisible again but this time more of the page was rendered into the DOM and a matching element was appended before the element we matched at first, that element became visible and, thus, satisfies the locator requirements. 

`operation` here means anything we do internally inside locator that can timeout. `timeout` is the user's timeout for the entire locator action.

Reply:
Discussed offline.